core/primitives/src/time.rs

```diff
@@ -0,0 +1,156 @@
+use std::default::Default;
+use std::sync::{Arc, Mutex};
+
+use chrono;
+use once_cell::sync::Lazy;
+
+pub use chrono::Utc;
+pub use std::time::{Duration, Instant};
+
+use chrono::DateTime;
+use std::collections::{HashMap, VecDeque};
+use std::ops::Deref;
+use std::thread::ThreadId;
+
+pub struct MockClockPerThread {
+    utc: VecDeque<DateTime<Utc>>,
+    durations: VecDeque<Duration>,
+    utc_call_count: u64,
+    instant_call_count: u64,
+    instant: Instant,
+    is_mock: bool,
+}
+
+pub struct Clock {
+    clocks_per_thread: HashMap<ThreadId, MockClockPerThread>,
+}
+
+impl MockClockPerThread {
+    pub fn reset(&mut self) {
+        self.utc.clear();
+        self.durations.clear();
+        self.utc_call_count = 0;
+        self.instant_call_count = 0;
+        self.instant = Instant::now();
+        self.is_mock = false;
+    }
+}
+
+impl Default for MockClockPerThread {
+    fn default() -> Self {
+        Self {
+            utc: VecDeque::with_capacity(16),
+            durations: VecDeque::with_capacity(16),
+            utc_call_count: 0,
+            instant_call_count: 0,
+            instant: Instant::now(),
+            is_mock: false,
+        }
+    }
+}
+
+impl Default for Clock {
+    fn default() -> Self {
+        Self { clocks_per_thread: HashMap::default() }
+    }
+}
+
+static SINGLETON: Lazy<Arc<Mutex<Clock>>> = Lazy::new(|| Arc::from(Mutex::new(Clock::new())));
+
+impl Clock {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn get() -> &'static Arc<Mutex<Clock>> {
+        SINGLETON.deref()
+    }
```

Comment:
```suggestion
    pub fn get() -> MutexGuard<'static, Clock> {
        SINGLETON.lock().unwrap()
    }
```

would simplify call-sites.